python/src/count_sketch.py

```diff
@@ -0,0 +1,240 @@
+            a_sign, b_sign = self.sign_hash_params[ii]  # Gets the (a,b) pair used for *sign* hashes at level ii
+            bucket = self._bucket_hash(item, a_bucket, b_bucket, self.w)
+            sign = self._sign_hash(item, a_sign, b_sign)
+            self.table[ii, bucket] += sign*weight
+
+    def update(self, item: int, weight=1.0):
+        """
+        Updates the sketch by:
+         1. Inserting (item, weight) into self.table
+         2. Estimating the current frequency and pushing it into the heavy_hitter_detector
+         (nb. item only pushed into the detector if it has large enough frequency.
+         This logic is dealt with in the detector class itself.)
+        :param item:int
+        :param weight:float
+        """
+        self.total_weight += weight
+        self._insert(item, weight)
+        self.heavy_hitter_detector.push(item, self.get_estimate(item))
+
+    def get_epsilon(self):
+        """
+        :return: an estimate of the worst-case epsilon error achieved by the sketch for frequency estimation.
+        """
+        return np.sqrt(np.e / self.w)
+
+    def get_estimate(self, item: int):
+        """
+        :param item:int -- the identifier of the item whose frequency is being queried.
+        :return:np.median(_estimates) -- The count sketch frequency estimate.
+        """
+        _estimates = self._get_frequency_estimate(item)
+        for sk in self.merged_sketches:
+            _estimates += sk._get_frequency_estimate(item)
+        return np.median(_estimates)  # TODO - replace this with np.max(0.0, np.median(_estimate))?
+
+    def _get_frequency_estimate(self, item: int):
+        """
+        :param item:
+        :return: _estimates -- an array of the frequency estimate where each entry corresponds to a level of the sketch.
+        """
+        _buckets = {_: self._bucket_hash(item, self.bucket_hash_params[_][0], self.bucket_hash_params[_][1], self.w)
+                    for _ in range(self.d)}
+        _signs = {_: self._sign_hash(item, self.sign_hash_params[_][0], self.sign_hash_params[_][1])
+                  for _ in range(self.d)}
+        _estimates = np.array([self.table[_, _buckets[_]] * _signs[_] for _ in range(self.d)])
+        return _estimates
+
+    def get_frequent_items(self):
+        """
+        :return: a list of the heavy hitters and their frequencies.
+        """
+        for k in self.heavy_hitter_detector.heap.keys():
+            self.heavy_hitter_detector.heap[k] = self.get_estimate(k)
+        return list(self.heavy_hitter_detector.heap.items())
+
+    def get_total_weight(self):
+        """
+        Returns the total weight inserted on the stream.
+        Is equivalent to the mass inserted over the stream, or the ell_1 norm of the
+        underlying frequency vector.
+        """
+        return self.total_weight
+
+    def is_empty(self):
+        """
+        Returns True if the sketch self.table is empty and is False otherwise.
```

Comment:
There's a valid philosophical debate over whether adding and then subtracting an item means a sketch is truly empty. It might seem trivial with a single item, but imagine merging 2 sketches that were created with the same items but opposite weights -- if we treat it as empty, we lose the information that it has ingested data.

Certainly open to the idea that we don't care about a distinction, but it's worth thinking about what we want to capture by "empty"